libs/langchain/langchain/document_loaders/recursive_url_loader.py

```diff
@@ -15,173 +54,106 @@ class RecursiveUrlLoader(BaseLoader):
+        self.timeout = timeout
+        self.prevent_outside = prevent_outside if prevent_outside is not None else True
+        self.link_regex = link_regex
+        self._lock = asyncio.Lock() if self.use_async else None
+        self.headers = headers
 
     def _get_child_links_recursive(
-        self, url: str, visited: Optional[Set[str]] = None, depth: int = 0
+        self, url: str, visited: Set[str], *, depth: int = 0
     ) -> Iterator[Document]:
         """Recursively get all child links starting with the path of the input URL.
 
         Args:
             url: The URL to crawl.
             visited: A set of visited URLs.
+            depth: Current depth of recursion. Stop when depth >= max_depth.
         """
 
-        if depth > self.max_depth:
-            return []
-
-        # Add a trailing slash if not present
-        if not url.endswith("/"):
-            url += "/"
-
-        # Exclude the root and parent from a list
-        visited = set() if visited is None else visited
-
+        if depth >= self.max_depth:
+            return
         # Exclude the links that start with any of the excluded directories
-        if self.exclude_dirs and any(
-            url.startswith(exclude_dir) for exclude_dir in self.exclude_dirs
-        ):
-            return []
+        if any(url.startswith(exclude_dir) for exclude_dir in self.exclude_dirs):
+            return
 
         # Get all links that can be accessed from the current URL
         try:
-            response = requests.get(url, timeout=self.timeout)
+            response = requests.get(url, timeout=self.timeout, headers=self.headers)
         except Exception:
-            return []
-
-        absolute_paths = self._get_sub_links(response.text, url)
+            logger.warning(f"Unable to load from {url}")
+            return
+        content = self.extractor(response.text)
+        if content:
+            yield Document(
+                page_content=content,
+                metadata=self.metadata_extractor(response.text, url),
+            )
+        visited.add(url)
 
         # Store the visited links and recursively visit the children
-        for link in absolute_paths:
+        sub_links = extract_sub_links(
+            response.text,
+            self.url,
+            pattern=self.link_regex,
+            prevent_outside=self.prevent_outside,
+        )
+        for link in sub_links:
             # Check all unvisited links
             if link not in visited:
-                visited.add(link)
-
-                try:
-                    response = requests.get(link)
-                    text = response.text
-                except Exception:
-                    # unreachable link, so just ignore it
-                    continue
-                loaded_link = Document(
-                    page_content=self.extractor(text),
-                    metadata=self._gen_metadata(text, link),
+                yield from self._get_child_links_recursive(
```

Comment:
this only works becasue `visited` is mutated... imo thats a little sneaky and not super apparent at all to readers. is there a way not rely on this mutation? it wasnt like this before right?

Reply:
think this is how it was, see L180 previously

Reply:
are you suggesting having a helper that explicitly returns visited?

Reply:
oh youre right